sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1484,6 +1511,26 @@ class SessionCatalog(
       require(functionBuilder.isDefined, s"built-in function '$f' is missing function builder")
       functionRegistry.registerFunction(f, expressionInfo.get, functionBuilder.get)
     }
+    invalidateAllCachedCatalogTable()
+  }
+
+  private[sql] def getCachedCatalogTable(qtn: QualifiedTableName): Option[CatalogTable] = {
+    cachedCatalogTable.getIfPresent(qtn) match {
+      case null => None
+      case catalogTable => Some(catalogTable)
+    }
+  }
+
+  private[sql] def cacheCatalogTable(qtn: QualifiedTableName, catalogTable: CatalogTable): Unit = {
```

Comment:
Caches should pass a Callable so that populating the cache can be combined with a get operation (get or initialize).

Instead of `cacheCatalogTable`, this should be `getOrCacheCatalogTable(qtn: QualifiedTableName, init: Callable[CatalogTable])` that calls `catalogTableCache.get(qtn, init)`.

Reply:
OK, fix these.

Reply:
Here `getOrCacheCatalogTable` is just a simple wrapper doing nothing.
I suggest to move the lambda parameter in the caller into this method. It is in fact a part of "Get or Cache".